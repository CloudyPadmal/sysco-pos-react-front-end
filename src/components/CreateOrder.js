import React, { Component } from 'react';
import { withRouter, Redirect } from 'react-router-dom'
import ListItemInEditOrder from './ListItemInEditOrder';

/**
 * @abstract Creates a new order
 * @description Creates a new order with multiple items added to it
 * @param ISLOGGEDIN
 * @param CURRENTORDERID
 * @param ITEMSLIST
 * @param DELETE_THIS_ORDER
 * @param CREATE_NEW_ORDER_FOR_THIS_USER
 * @param ADD_ITEMS_TO_THIS_ORDER
 */
class CreateOrder extends Component {

    constructor(props) {
        super(props);
        this.state = {
            TOTAL: 0
        };
    }

    // As the create order view mounted, generate a new order for the user
    // and in the meantime fetch the complete item list
    componentDidMount() {
        this.props.CREATE_NEW_ORDER_FOR_THIS_USER();
    }

    /**************************************************************************
     * Once all the items are set, this method will trigger the order state 
     * updating function in parent component to trigger a series of axios 
     * requests adding the items selected to the order and update the global 
     * item count. Before all that, this will filter the items for zero item 
     * counts. Once the parent function is triggered, redirects user to order
     * list after waiting for a time out depending on the number of items being
     * added to show user the completed order rather than a half filled order
     * as it takes some time to complete the axios promises.
     *************************************************************************/
    ADD_THESE_ITEMS_TO_THIS_ORDER = () => {
        this.props.ADD_ITEMS_TO_THIS_ORDER();
        setTimeout(() => {
            this.props.history.push('/my_orders')
        }, (50 * this.props.ITEMQUANTITY.length));
    }

    /**************************************************************************
     * If user chose not to create this order, this local function will trigger
     * parent component function to delete the current order being created. 
     * Once done, user will be redirected to order list.
     *************************************************************************/
    CANCEL_THE_ORDER = (ID) => {
        this.props.DELETE_THIS_ORDER(ID);
        this.props.CLEAR_ORDER_ADDING_PROCESS();
        this.props.history.push('/my_orders');
    }

    render() {
        if (!this.props.ISLOGGEDIN) {
            return (
                <Redirect to="/login" />
            );
        } else {
            return (
                <div>
                    <div className="card" style={{ margin: '25px', paddingBottom: '50px' }}>
                        <div className="card-body">
                            <h5 className="card-title"
                                style={{ margin: '0.5em 1em' }}>Item List</h5>
                            {this.props.ITEMSLIST.map((item) => (
                                <ListItemInEditOrder
                                    key={item._id}
                                    ITEM={item}
                                    NAME={item.productID}
                                    ITEMQUANTITY={this.props.ITEMQUANTITY[item.productID] === undefined
                                        ? 0 : this.props.ITEMQUANTITY[item.productID]}
                                    ADD_THIS_ITEM_TO_ITEMQUANTITY={this.props.ADD_THIS_ITEM_TO_ITEMQUANTITY}
                                    DELETE_THIS_ITEM={this.props.DELETE_THIS_ITEM}
                                    INDECCREMENT_ITEM_COUNT={this.props.INDECCREMENT_ITEM_COUNT} />
                            ))}
                        </div>
                    </div>
                    <nav className="navbar navbar-light bg-light"
                        style={{
                            paddingRight: '25px',
                            overflow: 'hidden',
                            position: 'fixed',
                            zIndex: 99,
                            bottom: '0',
                            width: '100%'
                        }}>
                        <div className="card-body" style={{ paddingTop: '0px' }}>
                            <div className="row">
                                <div className="col-12 d-flex justify-content-end">
                                    <button
                                        onClick={this.ADD_THESE_ITEMS_TO_THIS_ORDER}
                                        className="btn btn-success"
                                        style={{ marginRight: '10px' }}>Add to Cart</button>
                                    <button
                                        onClick={this.CANCEL_THE_ORDER.bind(
                                            this, this.props.CURRENTORDERID
                                        )}
                                        className="btn btn-danger">Cancel</button>
                                </div>
                            </div>
                        </div>
                    </nav>
                </div>
            );
        }
    }
}

export default withRouter(CreateOrder);
