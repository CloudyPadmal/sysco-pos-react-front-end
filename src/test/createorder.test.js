import React from 'react';
import { shallowWrap, mountWrap } from './enzyme';

import CreateOrder from '../components/CreateOrder';

describe('Create Order', () => {

    it('displays a list of items to create order', () => {
        const wrapper = mountWrap(<CreateOrder
            ISLOGGEDIN={true}
            CURRENTORDERID={12441232}
            ITEMSLIST={[]}
            ITEMQUANTITY={{}}
            TOTAL={14234.33}
            DELETE_THIS_ORDER={() => { }}
            DELETE_THIS_ITEM={() => { }}
            INDECCREMENT_ITEM_COUNT={() => { }}
            CREATE_NEW_ORDER_FOR_THIS_USER={() => { }}
            CLEAR_ORDER_ADDING_PROCESS={() => { }}
            ADD_THIS_ITEM_TO_ITEMQUANTITY={() => { }}
            ADD_ITEMS_TO_THIS_ORDER={() => { }} />);
        expect(wrapper).toMatchSnapshot();
    });

    it('redirects to login page if not logged in', () => {
        const wrapper = shallowWrap(<CreateOrder
            ISLOGGEDIN={false}
            CURRENTORDERID={12441232}
            ITEMSLIST={[]}
            ITEMQUANTITY={{}}
            TOTAL={14234.33}
            DELETE_THIS_ORDER={() => { }}
            DELETE_THIS_ITEM={() => { }}
            INDECCREMENT_ITEM_COUNT={() => { }}
            CREATE_NEW_ORDER_FOR_THIS_USER={() => { }}
            CLEAR_ORDER_ADDING_PROCESS={() => { }}
            ADD_THIS_ITEM_TO_ITEMQUANTITY={() => { }}
            ADD_ITEMS_TO_THIS_ORDER={() => { }} />);
        expect(wrapper).toMatchSnapshot();
    });
});