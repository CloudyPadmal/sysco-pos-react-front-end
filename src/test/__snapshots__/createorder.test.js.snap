// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Create Order displays a list of items to create order 1`] = `
<withRouter(CreateOrder)
  ADD_ITEMS_TO_THIS_ORDER={[Function]}
  ADD_THIS_ITEM_TO_ITEMQUANTITY={[Function]}
  CLEAR_ORDER_ADDING_PROCESS={[Function]}
  CREATE_NEW_ORDER_FOR_THIS_USER={[Function]}
  CURRENTORDERID={12441232}
  DELETE_THIS_ITEM={[Function]}
  DELETE_THIS_ORDER={[Function]}
  INDECCREMENT_ITEM_COUNT={[Function]}
  ISLOGGEDIN={true}
  ITEMQUANTITY={Object {}}
  ITEMSLIST={Array []}
  TOTAL={14234.33}
>
  <Route>
    <CreateOrder
      ADD_ITEMS_TO_THIS_ORDER={[Function]}
      ADD_THIS_ITEM_TO_ITEMQUANTITY={[Function]}
      CLEAR_ORDER_ADDING_PROCESS={[Function]}
      CREATE_NEW_ORDER_FOR_THIS_USER={[Function]}
      CURRENTORDERID={12441232}
      DELETE_THIS_ITEM={[Function]}
      DELETE_THIS_ORDER={[Function]}
      INDECCREMENT_ITEM_COUNT={[Function]}
      ISLOGGEDIN={true}
      ITEMQUANTITY={Object {}}
      ITEMSLIST={Array []}
      TOTAL={14234.33}
      history={
        Object {
          "action": "POP",
          "block": [Function],
          "createHref": [Function],
          "go": [Function],
          "goBack": [Function],
          "goForward": [Function],
          "length": 1,
          "listen": [Function],
          "location": Object {
            "hash": "",
            "pathname": "/",
            "search": "",
            "state": undefined,
          },
          "push": [Function],
          "replace": [Function],
        }
      }
      location={Object {}}
      match={Object {}}
    >
      <div>
        <div
          className="card"
          style={
            Object {
              "margin": "25px",
              "paddingBottom": "50px",
            }
          }
        >
          <div
            className="card-body"
          >
            <h5
              className="card-title"
              style={
                Object {
                  "margin": "0.5em 1em",
                }
              }
            >
              Item List
            </h5>
          </div>
        </div>
        <nav
          className="navbar navbar-light bg-light"
          style={
            Object {
              "bottom": "0",
              "overflow": "hidden",
              "paddingRight": "25px",
              "position": "fixed",
              "width": "100%",
              "zIndex": 99,
            }
          }
        >
          <div
            className="card-body"
            style={
              Object {
                "paddingTop": "0px",
              }
            }
          >
            <div
              className="row"
            >
              <div
                className="col-4 d-flex justify-content-start d-inline"
              >
                <div
                  className="input-group-prepend"
                  style={
                    Object {
                      "marginLeft": "10px",
                    }
                  }
                >
                  <span
                    className="input-group-text"
                    style={
                      Object {
                        "borderBottomRightRadius": "0px",
                        "borderTopRightRadius": "0px",
                      }
                    }
                  >
                    <strong>
                      Total
                    </strong>
                  </span>
                  <span
                    className="input-group-text"
                    style={
                      Object {
                        "borderBottomLeftRadius": "0px",
                        "borderTopLeftRadius": "0px",
                      }
                    }
                  >
                    Rs. 
                    14234.33
                  </span>
                </div>
              </div>
              <div
                className="col-8 d-flex justify-content-end"
              >
                <button
                  className="btn btn-success"
                  onClick={[Function]}
                  style={
                    Object {
                      "marginRight": "10px",
                    }
                  }
                >
                  <i
                    className="fas fa-cart-plus"
                  />
                   Add to Cart
                </button>
                <button
                  className="btn btn-danger"
                  onClick={[Function]}
                >
                  <i
                    className="fas fa-times-circle"
                  />
                   Cancel
                </button>
              </div>
            </div>
          </div>
        </nav>
      </div>
    </CreateOrder>
  </Route>
</withRouter(CreateOrder)>
`;

exports[`Create Order redirects to login page if not logged in 1`] = `
<Route>
  <Component />
</Route>
`;
